### PR TITLE
Use numerically stable functions from statsfuns

### PR DESCRIPTION
**Pull request by mihirparadkar**
I'm opening this PR for discussion purposes. StatsFuns has quite a few useful functions for probabilistic loss functions (cross-entropy, logistic margin, logitdistance, etc.), and they emphasize numerical stability (currently, `value(LogitDistLoss(), 1000)` gives `Inf` when it should give `1000`, but it exhibits proper behavior when using softplus.

### CODE REVIEW

**Comment by joshday** [comment]
StatsFuns.logistic is already what we are using, just not wrapped in a function:
https://github.com/JuliaStats/StatsFuns.jl/blob/c80fc2c4dfe7873a8e34850af2a280e890c13f2f/src/basicfuns.jl#L12

softmax = log1pexp is only two lines:
https://github.com/JuliaStats/StatsFuns.jl/blob/c80fc2c4dfe7873a8e34850af2a280e890c13f2f/src/basicfuns.jl#L23

**Picking up StatsFuns also adds Compat, RMath, and SpecialFunctions**, which is a bit much.

I say we copy those two lines or use the log-sum-exp trick here: https://www.xarg.org/2016/06/the-log-sum-exp-trick-in-machine-learning/

Something like:
```julia
julia> function log1pexp(x)
           x > 0 ? x + log(1 + exp(-x)) : log(1 + exp(x))
       end
log1pexp (generic function with 1 method)

julia> log1pexp(1000.)
1000.0
```

**Comment by joshday** [comment]
Also, good find on the numerical stability 👍

**Comment by mihirparadkar** [comment]
Yeah, I agree that the dependencies (particularly RMath) are a bit heavy for StatsFuns, unfortunately. 

I am aware of the logsumexp trick (LowRankModels and StatsFuns both use it for stability), and log1pexp is not difficult to implement. Perhaps splitting off the non RMath-dependent functions (everything seen here and more potentially useful ones) into another package is something to discuss with the authors. Perhaps I'm being idealistic, but I think it's still best if these functions, however small, lived in one place.

**Comment by joshday** [comment]
My vote would be to borrow code as a temporary fix, with a comment that it comes from StatsFuns.

>  I think it's still best if these functions, however small, lived in one place.

Agreed, and LossFunctions could probably make use of other things in StatsFuns, but splitting off the Rmath-required stuff would then affect Distributions and possibly other packages.  I'm unsure of the amount of effort that would take.

**Comment by mihirparadkar** [comment]
Borrowing code is probably best for now, especially since these particular functions are so simple to implement. It also may be non-trivial to split StatsFuns due to dependencies, but I guess that's the benefit of a language that isn't at 1.0 yet (can't hurt to ask).
